resources/js/Pages/Galleries/hooks/useGalleryDrafts.ts

```diff
@@ -0,0 +1,142 @@
+    coverType: string | null;
+    nfts: DraftNft[];
+    walletAddress?: string;
+    id: number | null;
+}
+
+const initialGalleryDraft: GalleryDraft = {
+    title: "",
+    cover: null,
+    coverType: null,
+    nfts: [],
+    id: null,
+};
+
+const MAX_DRAFT_LIMIT_PER_WALLET = 6;
+
+interface GalleryDraftsState {
+    reachedLimit: boolean;
+    draft: GalleryDraft;
+    setDraftCover: (image: ArrayBuffer | null, type: string | null) => void;
+    setDraftNfts: (nfts: App.Data.Gallery.GalleryNftData[]) => void;
+    setDraftTitle: (title: string) => void;
+}
+
+export const useGalleryDrafts = (givenDraftId?: number): GalleryDraftsState => {
+    const { wallet } = useAuth();
+
+    const database = useIndexedDB("gallery-drafts");
+
+    const [draft, setDraft] = useState<GalleryDraft>({
+        ...initialGalleryDraft,
+        walletAddress: wallet?.address,
+    });
+
+    const [save, setSave] = useState(false);
+    const [isSaving, setIsSaving] = useState(false);
+
+    const [title, setTitle] = useState<string>("");
+    const [debouncedValue] = useDebounce(title, 400);
+
+    const isFirstRender = useIsFirstRender();
+
+    const [reachedLimit, setReachedLimit] = useState(false);
+
+    // populate `draft` state if `givenDraftId` is present
+    useEffect(() => {
+        if (givenDraftId === undefined) return;
+        const getDraft = async (): Promise<void> => {
+            const draft: GalleryDraft = await database.getByID(givenDraftId);
+
+            if (draft.walletAddress === wallet?.address) {
+                setDraft(draft);
+            }
+        };
+
+        void getDraft();
+    }, []);
+
+    // handle debounced title
+    useEffect(() => {
+        if (isFirstRender) return;
+
+        setDraft({ ...draft, title: debouncedValue });
+        setSave(true);
+    }, [debouncedValue]);
+
+    useEffect(() => {
+        if (!save || isSaving || reachedLimit) return;
+
+        void saveDraft();
+    }, [save]);
+
+    const saveDraft = async (): Promise<void> => {
+        setIsSaving(true);
+
+        if (draft.id === null) {
+            const walletDrafts = await getWalletDrafts();
+
+            if (walletDrafts.length >= MAX_DRAFT_LIMIT_PER_WALLET) {
+                setReachedLimit(true);
+                return;
+            }
+
+            const draftToCreate: Partial<GalleryDraft> = { ...draft };
+            delete draftToCreate.id;
+
+            const id = await database.add(draftToCreate);
+            setDraft({ ...draft, id });
+        } else {
+            await database.update(draft);
+        }
+
+        setSave(false);
+        setIsSaving(false);
+    };
+
+    const getWalletDrafts = async (): Promise<GalleryDraft[]> => {
+        const allDrafts: GalleryDraft[] = await database.getAll();
+
+        return allDrafts.filter((draft) => draft.walletAddress === wallet?.address);
+    };
+
+    const setDraftCover = (image: ArrayBuffer | null, type: string | null): void => {
```

Comment:
We have to use `ArrayBuffer` to store images as Safari cannot store `Blob`s

> If you are storing large, user-generated files such as images or videos, then you may try to store them as File or Blob objects. This will work on some platforms but fail on others. Safari on iOS, in particular, cannot store Blobs in IndexedDB.

> Luckily it is not too difficult to convert a Blob into an ArrayBuffer, and vice versa. Storing ArrayBuffers in IndexedDB is very well supported.

> Remember, however, that a Blob has a MIME type while an ArrayBuffer does not. You will need to store the type alongside the buffer in order to do the conversion correctly.

https://web.dev/articles/indexeddb-best-practices